[cmake] Fix recompilation on rebuild without changes

ntrrgc opened the pull request:
#### 18c724980da23899f80a0271c0cbb70c0769e662
<pre>
[cmake] Fix recompilation on rebuild without changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=257516">https://bugs.webkit.org/show_bug.cgi?id=257516</a>

Reviewed by Adrian Perez de Castro.

Running build-webkit twice in a row made the second run still take
somewhere in the order of a minute to complete, despite there being no
changes in any code.

This patch fixes the bugs in the build system causing that deficiency,
at least for users of CMake &gt;=3.20.

Namely:

* Amends 250761@main, which added a cmake_policy() command for CMP0116
  without any effect. This patch sets it to NEW, which tells CMake to
  rewrite depfiles to use paths relative to the sub-build directory if
  the user CMake version is recent enough (&gt;=3.20). Previous versions of
  CMake will remain affected.

* Refactors all usages of glib-compile-commands into a new CMake
  function, GLIB_COMPILE_COMMANDS(), provided by a new
  FindGLibCompileResources.cmake file.

* Detects versions of the glib-compile-commands older than 2.77 as
  potentially affected by this bug in depfile generation:
  <a href="https://gitlab.gnome.org/GNOME/glib/-/issues/2829">https://gitlab.gnome.org/GNOME/glib/-/issues/2829</a>
  If building with an affected version, a Perl post processing script is
  added automatically that corrects the depfiles. The Perl script is
  meant to be harmless in the case of false positives (older but patched
  versions).

* Modifies apply-build-revision-to-files.py, which is responsible for
  generating BuildRevision.h, to not write to the file if the resulting
  contents would remain unchanged. This avoids touching the modification
  time of the file which build systems like Ninja use to test whether
  files are dirty and therefore need to be recompiled along with all
  their dependents. A similar check already exists for many other
  generators that run on every build.

* CMakeLists.txt:
* Source/WebKit/InspectorGResources.cmake:
* Source/WebKit/ModernMediaControlsGResources.cmake:
* Source/WebKit/PdfJSGResources.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/cmake/FindGLibCompileResources.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/glib/apply-build-revision-to-files.py:
(main):
* Tools/glib/fix-glib-resources-depfile.pl: Added.

Canonical link: <a href="https://commits.webkit.org/277059@main">https://commits.webkit.org/277059@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d872866fb32d15c512ed30a6fdf30ebced496fd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46572 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25729 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49179 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49249 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42614 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48879 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30089 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23191 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47151 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/22722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19242 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41249 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4617 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/39819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41629 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51097 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46053 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21579 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/17986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10291 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53197 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22572 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/10910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/InspectorGResources.cmake:
```diff
@@ -8,11 +8,10 @@ macro(WEBKIT_BUILD_INSPECTOR_GRESOURCES _derived_sources_dir)
+        OUTPUT        ${_derived_sources_dir}/InspectorGResourceBundle.c
+        DEPFILE       ${_derived_sources_dir}/InspectorGResourceBundle.deps
+        SOURCE_XML    ${_derived_sources_dir}/InspectorGResourceBundle.xml
+        RESOURCE_DIRS ${_derived_sources_dir}/InspectorResources/WebInspectorUI
```
I'm somewhat surprised that the style checker did not bark at the spaces between the keywords and the arguments. Now that I know it won't, probably I will be doing the same myself in the future&mdash;this kind of alignment makes some of these long invocations easier to follow for me!

aperezdc commented on Source/cmake/FindGLibCompileResources.cmake:
```diff
@@ -0,0 +1,84 @@
+message(STATUS "Found glib-compile-resources (required): "
+    ${GLIB_COMPILE_RESOURCES_EXE}
+    " (${glib_compile_resources_version})")
+if (${glib_compile_resources_version} VERSION_LESS 2.77)
```
I would quote the expansion of the variable, just in case something bad happens during execution and it ends up being empty, and that way avoid a syntax error from CMake:

```suggestion
if ("${glib_compile_resources_version}" VERSION_LESS 2.77)
```

aperezdc commented on Source/cmake/FindGLibCompileResources.cmake:
```diff
@@ -0,0 +1,84 @@
+    message(STATUS "  Needs workaround for depfiles (will be used automatically)")
+else ()
+    set(glib_compile_resources_has_depfile_bug FALSE)
+    message(STATUS "  Doesn't need workarounds")
```
I think we do not need to inform about this, no need to make the find-module chatty.

```suggestion
else ()
    set(glib_compile_resources_has_depfile_bug FALSE)
```

aperezdc commented on Source/cmake/FindGLibCompileResources.cmake:
```diff
@@ -0,0 +1,84 @@
+            }
+            s{\Q$old_xml:}{$new_target: $old_xml}
+        ]==])
+        file(WRITE ${CMAKE_BINARY_DIR}/fix_glib_compile_resources_depfile.pl ${perl_code})
```
This will re-write the `fix_glib_compile_resources_depfile.pl` on each invocation of the `GLIB_COMPILE_RESOURCES()` CMake function. Let's not do that; this should go into `Scripts/Tools/fix-glib-resources-depfile.pl`.

aperezdc commented on Source/cmake/FindGLibCompileResources.cmake:
```diff
@@ -0,0 +1,84 @@
+
+function(GLIB_COMPILE_RESOURCES)
+    set(zeroArgKeywords "")
+    set(oneArgKeywords OUTPUT SOURCE_XML DEPFILE)
```
From what I can see there is no need to have a `DEPFILE` keyword argument. This function should instead add a dependency file automatically:

```suggestion
    set(oneArgKeywords OUTPUT SOURCE_XML)
```

The path of the dependency file can be derived from the output file path, see my suggestion below.

aperezdc commented on Source/cmake/FindGLibCompileResources.cmake:
```diff
@@ -0,0 +1,84 @@
+        COMMAND ${GLIB_COMPILE_RESOURCES_EXE}
+                --generate
+                --target=${rel_output}
+                --dependency-file=${ARG_DEPFILE}
```
To accomodate for the removed `DEPFILE` keyword argument, derive the path from `ARG_OUTPUT`, for example:

```suggestion
        OUTPUT  ${ARG_OUTPUT} ${ARG_OUTPUT}.deps
        DEPENDS ${ARG_SOURCE_XML}
        DEPFILE ${ARG_OUTPUT}.deps
        COMMAND ${GLIB_COMPILE_RESOURCES_EXE}
                --generate
                --target=${rel_output}
                --dependency-file=${ARG_OUTPUT}.deps
```

aperezdc commented on Source/WebKit/InspectorGResources.cmake:
```diff
@@ -8,11 +8,10 @@ macro(WEBKIT_BUILD_INSPECTOR_GRESOURCES _derived_sources_dir)
-        VERBATIM
+    GLIB_COMPILE_RESOURCES(
+        OUTPUT        ${_derived_sources_dir}/InspectorGResourceBundle.c
+        DEPFILE       ${_derived_sources_dir}/InspectorGResourceBundle.deps
```
I think we don't need to specify `DEPFILE` on each call to the `GLIB_COMPILE_RESOURCES()` function, please read on below for an alternative.

aperezdc reviewed (changes requested):
@ntrrgc Thanks for working on this! I have tested this locally and indeed greatly improves the speed of incremental builds 🚀 

Please take a look at the comments, it's mostly small things, and then in particular removing the need to specify `DEPFILE` would make it easier to use the new CMake function. I'll be more than happy to approve this afterwards.

aperezdc reviewed (approved):
Thanks for the update @ntrrgc, looking very nice now! 👍🏼

aperezdc commented on Source/cmake/FindGLibCompileResources.cmake:
```diff
@@ -0,0 +1,68 @@
+        # But depfiles should look like this:
+        #   foo.c: foo.xml resource1 resource2
+        set(additional_cmd_line &&
+            perl -pi ${CMAKE_SOURCE_DIR}/Tools/Scripts/fix-glib-resources-depfile.pl
```
There is also a `${TOOLS_DIR}` variable, which could be used instead of `${CMAKE_SOURCE_DIR}/Tools`.

aperezdc commented on Tools/glib/apply-build-revision-to-files.py:
```diff
@@ -76,13 +76,20 @@ def main(args):
+        # Only write the new contents to the output file if there are any changes.
+        # Otherwise, the file would become 'dirty' and all dependents would be
+        # recompiled unnecessarily.
+        with file.open() as fd:
+            old_contents = fd.read()
+        if old_contents != new_contents:
+            with file.open('w') as fd:
+                fd.write(new_contents)
```
Nice, good catch!

ntrrgc replied:
I just realized this will fail if the header file doesn't exist when run.

ntrrgc replied:
nvm, this will not occur because the file is first created as just a copy of the template by cmake.

```cmake
configure_file(Shared/glib/BuildRevision.h.in ${FORWARDING_HEADERS_WPE_DIR}/BuildRevision.h)
```

aperezdc reviewed (commented):
LGTM, let's land this 💯

ntrrgc committed:
[cmake] Fix recompilation on rebuild without changes
https://bugs.webkit.org/show_bug.cgi?id=257516

Reviewed by Adrian Perez de Castro.

Running build-webkit twice in a row made the second run still take
somewhere in the order of a minute to complete, despite there being no
changes in any code.

This patch fixes the bugs in the build system causing that deficiency,
at least for users of CMake >=3.20.

Namely:

* Amends 250761@main, which added a cmake_policy() command for CMP0116
  without any effect. This patch sets it to NEW, which tells CMake to
  rewrite depfiles to use paths relative to the sub-build directory if
  the user CMake version is recent enough (>=3.20). Previous versions of
  CMake will remain affected.

* Refactors all usages of glib-compile-commands into a new CMake
  function, GLIB_COMPILE_COMMANDS(), provided by a new
  FindGLibCompileResources.cmake file.

* Detects versions of the glib-compile-commands older than 2.77 as
  potentially affected by this bug in depfile generation:
  https://gitlab.gnome.org/GNOME/glib/-/issues/2829
  If building with an affected version, a Perl post processing script is
  added automatically that corrects the depfiles. The Perl script is
  meant to be harmless in the case of false positives (older but patched
  versions).

* Modifies apply-build-revision-to-files.py, which is responsible for
  generating BuildRevision.h, to not write to the file if the resulting
  contents would remain unchanged. This avoids touching the modification
  time of the file which build systems like Ninja use to test whether
  files are dirty and therefore need to be recompiled along with all
  their dependents. A similar check already exists for many other
  generators that run on every build.

* CMakeLists.txt:
* Source/WebKit/InspectorGResources.cmake:
* Source/WebKit/ModernMediaControlsGResources.cmake:
* Source/WebKit/PdfJSGResources.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/cmake/FindGLibCompileResources.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/glib/apply-build-revision-to-files.py:
(main):
* Tools/glib/fix-glib-resources-depfile.pl: Added.

Canonical link: https://commits.webkit.org/277059@main
